File: app/src/main/java/cn/surine/schedulex/ui/schedule_config/ScheduleConfigFragment.java
package cn.surine.schedulex.ui.schedule_config;

import android.Manifest;
import android.annotation.SuppressLint;
import android.content.ContentResolver;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.provider.MediaStore;
import android.text.TextUtils;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.SeekBar;
import android.widget.TextView;

import androidx.annotation.Nullable;
import androidx.lifecycle.ViewModelProviders;
import androidx.navigation.Navigation;
import androidx.navigation.fragment.NavHostFragment;

import com.bumptech.glide.Glide;
import com.google.android.material.bottomsheet.BottomSheetDialog;
import com.tbruyelle.rxpermissions2.RxPermissions;

import java.io.File;
import java.io.FileNotFoundException;

import cn.surine.schedulex.R;
import cn.surine.schedulex.base.Constants;
import cn.surine.schedulex.base.controller.App;
import cn.surine.schedulex.base.controller.BaseBindingFragment;
import cn.surine.schedulex.base.utils.Dates;
import cn.surine.schedulex.base.utils.Files;
import cn.surine.schedulex.base.utils.InstanceFactory;
import cn.surine.schedulex.base.utils.MySeekBarChangeListener;
import cn.surine.schedulex.base.utils.Navigations;
import cn.surine.schedulex.base.utils.Prefs;
import cn.surine.schedulex.base.utils.Toasts;
import cn.surine.schedulex.base.utils.Uis;
import cn.surine.schedulex.data.entity.Schedule;
import cn.surine.schedulex.databinding.FragmentScheduleConfigBinding;
import cn.surine.schedulex.ui.course.CourseRepository;
import cn.surine.schedulex.ui.course.CourseViewModel;
import cn.surine.schedulex.ui.schedule.ScheduleRepository;
import cn.surine.schedulex.ui.schedule.ScheduleViewModel;
import cn.surine.schedulex.ui.schedule_list.ScheduleListFragment;
import cn.surine.schedulex.ui.timetable_list.TimeTableRepository;
import cn.surine.schedulex.ui.timetable_list.TimeTableViewModel;
import cn.surine.schedulex.ui.view.custom.helper.CommonDialogs;

import static android.app.Activity.RESULT_OK;

/**
 * Intro：
 *
 * @author sunliwei
 * @date 2020-01-30 15:23
 */
@SuppressLint("SetTextI18n")
public class ScheduleConfigFragment extends BaseBindingFragment<FragmentScheduleConfigBinding> {

    public static final int CHANGE_SCHEDULE_NAME = 100;
    public static final int CHANGE_WEEK_INFO = 101;
    public static final int CHANGE_BACKGROUND = 102;
    public static final int CHANGE_COURSE_ITEM_HEIGHT = 103;
    private static final int PICK_PHOTO = 1;

    private ScheduleViewModel scheduleViewModel;
    int scheduleId;
    private CourseViewModel courseViewModel;
    private Schedule schedule;
    private FragmentScheduleConfigBinding globalT;
    private int mSettingItemTag;

    public static final String SCHEDULE_ID = "SCHEDULE_ID";
    private TimeTableViewModel timeTableViewModel;

    @Override
    public int layoutId() {
        return R.layout.fragment_schedule_config;
    }


    @SuppressLint("CheckResult")
    @Override
    protected void onInit(FragmentScheduleConfigBinding t) {
        globalT = t;
        Class[] classes = new Class[]{ScheduleRepository.class};
        Object[] args = new Object[]{ScheduleRepository.abt.getInstance()};
        scheduleViewModel = ViewModelProviders.of(this, InstanceFactory.getInstance(classes, args)).get(ScheduleViewModel.class);

        Class[] classesForCourse = new Class[]{CourseRepository.class};
        Object[] argsForCourse = new Object[]{CourseRepository.abt.getInstance()};
        courseViewModel = ViewModelProviders.of(this, InstanceFactory.getInstance(classesForCourse, argsForCourse)).get(CourseViewModel.class);

        Class[] classesForTimeTable = new Class[]{TimeTableRepository.class};
        Object[] argsForTimeTable = new Object[]{TimeTableRepository.abt.getInstance()};
        timeTableViewModel = ViewModelProviders.of(this, InstanceFactory.getInstance(classesForTimeTable, argsForTimeTable)).get(TimeTableViewModel.class);

        scheduleId = getArguments() == null ? -1 : getArguments().getInt(SCHEDULE_ID);

        if (scheduleId == Prefs.getLong(Constants.CUR_SCHEDULE, -1L)) {
            t.deleteSchedule.setVisibility(View.GONE);
        }

        schedule = scheduleViewModel.getScheduleById(scheduleId);
        t.setData(schedule);


        /***
         * 捷径
         */
        if (getArguments() != null && (mSettingItemTag = getArguments().getInt(ScheduleListFragment.FUNCTION_TAG, -1)) != -1) {
            switch (mSettingItemTag) {
                case CHANGE_WEEK_INFO:
                    showTimeConfigDialog();
                    break;
                case CHANGE_COURSE_ITEM_HEIGHT:
                    showItemHeightDialog();
                    break;
                case CHANGE_BACKGROUND:
                    chooseBackgroundPicture();
                    break;
                case CHANGE_SCHEDULE_NAME:
                    modifyScheduleName();
                    break;
                default:
                    break;
            }
        }


        t.deleteSchedule.setOnClickListener(v -> CommonDialogs.getCommonDialog(activity(), getString(R.string.warning), getString(R.string.delete_schedule_dialog_msg)
                , () -> {
                    //删除课表后需要删除所有相关课程
                    scheduleViewModel.deleteScheduleById(scheduleId);
                    courseViewModel.deleteCourseByScheduleId(scheduleId);
                    Toasts.toast(getString(R.string.schedule_is_delete));
                    NavHostFragment.findNavController(ScheduleConfigFragment.this).navigateUp();
                }, null).show());


        t.scheduleNameItem.setOnClickListener(v -> modifyScheduleName());
        t.scheduleWeekInfoItem.setOnClickListener(v -> showTimeConfigDialog());
        t.scheduleTimeTableItem.setOnClickListener(v -> {
            Bundle bundle = new Bundle();
            bundle.putLong(ScheduleConfigFragment.SCHEDULE_ID, scheduleId);
            Navigations.open(ScheduleConfigFragment.this, R.id.action_scheduleConfigFragment_to_timeTableListFragment, bundle);
        });
        try {
            t.textViewTimeTableSubtitle.setText(timeTableViewModel.getTimTableById(schedule.timeTableId).name);
        } catch (Exception e) {
            t.textViewTimeTableSubtitle.setText("No Data");
        }

        t.scheduleBackgroundItem.setOnClickListener(v -> chooseBackgroundPicture());
        t.scheduleBackgroundItem.setOnLongClickListener(v -> {
            schedule.imageUrl = "";
            scheduleViewModel.updateSchedule(schedule);
            Toasts.toast(App.context.getResources().getString(R.string.handle_success));
            return true;
        });
        if (!TextUtils.isEmpty(schedule.imageUrl)) {
            Glide.with(activity()).load(new File(schedule.imageUrl)).into(t.backgroundPic);
        }

        //设置颜色适配开关
        t.switchs.setChecked(schedule.lightText);
        t.schedulePaletteItem.setOnClickListener(v -> {
            schedule.lightText = !schedule.lightText;
            scheduleViewModel.updateSchedule(schedule);
            t.switchs.setChecked(schedule.lightText);
            t.paletteColorSubTitle.setText(schedule.lightText ? R.string.white_txt : R.string.black_txt);
        });

        //设置是否展示周末开关
        t.showWeekSwitchs.setChecked(schedule.isShowWeekend);
        t.scheduleShowWeekItem.setOnClickListener(v -> {
            schedule.isShowWeekend = !schedule.isShowWeekend;
            scheduleViewModel.updateSchedule(schedule);
            t.showWeekSwitchs.setChecked(schedule.isShowWeekend);
            t.showWeekSubTitle.setText(schedule.isShowWeekend ? R.string.show_weekend : R.string.not_show_weekend);
        });

        //打开透明度配置窗口
        t.scheduleCourseAlphaItem.setOnClickListener(v -> showCourseItemAlphaDialog());

        //配置最大节次
        t.scheduleMaxSessionItem.setOnClickListener(v -> showMaxSessionDialog());
        //配置课表最大高度
        t.scheduleCourseItemHeightItem.setOnClickListener(v -> showItemHeightDialog());

        //是否显示时间表
        t.showTimesSwitch.setChecked(schedule.isShowTime);
        t.scheduleShowTimeItem.setOnClickListener(v -> {
            schedule.isShowTime = !schedule.isShowTime;
            scheduleViewModel.updateSchedule(schedule);
            t.showTimesSwitch.setChecked(schedule.isShowTime);
            t.scheduleShowTimeItemSubTitle.setText(schedule.isShowTime ? R.string.show_time : R.string.not_show_timetable);
        });

        t.changeCourseCardUi.setOnClickListener(v -> {
            Bundle bundle = new Bundle();
            bundle.putInt(SCHEDULE_ID,scheduleId);
            Navigations.open(ScheduleConfigFragment.this,R.id.action_scheduleConfigFragment_to_paletteFragment,bundle);
        });

        Bundle bundle = new Bundle();
        bundle.putInt(SCHEDULE_ID, scheduleId);
        t.export.setOnClickListener(Navigation.createNavigateOnClickListener(R.id.action_scheduleConfigFragment_to_scheduleDataExport, bundle));

    }

    private void modifyScheduleName() {
        CommonDialogs.getEditDialog(activity(), schedule.name, true, s -> {
            schedule.name = s;
            globalT.courseNameSubTitle.setText(s);
            Toasts.toast(getString(R.string.update_success));
            scheduleViewModel.updateSchedule(schedule);
        }, null);
    }


    private void chooseBackgroundPicture() {
        RxPermissions rxPermissions = new RxPermissions(activity());
        rxPermissions.request(Manifest.permission.READ_EXTERNAL_STORAGE).subscribe(aBoolean -> {
            if (aBoolean) {
                startChoosePicture();
            } else {
                Toasts.toast(getString(R.string.permission_is_denied));
            }
        });
    }


    /**
     * 课表item高度
     */
    private void showItemHeightDialog() {
        BottomSheetDialog bt = new BottomSheetDialog(activity(), R.style.BottomSheetDialogTheme);
        View view;
        bt.setDismissWithAnimation(true);
        bt.setContentView(view = Uis.inflate(activity(), R.layout.view_schedule_time));
        bt.getWindow().findViewById(R.id.design_bottom_sheet).setBackgroundDrawable(new ColorDrawable(Color.TRANSPARENT));
        bt.show();
        view.animate().translationY(50);

        TextView tv = view.findViewById(R.id.dialog_title);
        SeekBar s1 = view.findViewById(R.id.seekBar);
        SeekBar s2 = view.findViewById(R.id.seekBar2);
        TextView t1 = view.findViewById(R.id.tvS1);
        TextView t2 = view.findViewById(R.id.tvS2);
        Button button = view.findViewById(R.id.button);

        tv.setText("配置课程格子高度");
        s2.setVisibility(View.GONE);
        t2.setVisibility(View.GONE);


        s1.setMax(150);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            s1.setMin(30);
        }
        s1.setProgress(schedule.itemHeight);
        t1.setText(schedule.itemHeight + "dp");
        s1.setOnSeekBarChangeListener(new MySeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                t1.setText(progress + "dp");
            }
        });
        button.setOnClickListener(v -> {
            if (s1.getProgress() < 30) {
                Toasts.toast(getString(R.string.cant_small_than_30));
            } else {
                schedule.itemHeight = s1.getProgress();
                globalT.scheduleCourseItemHeightSubtitle.setText(s1.getProgress() + "px");
                scheduleViewModel.updateSchedule(schedule);
                bt.dismiss();
            }
        });
    }


    /**
     * 最大节次
     */
    private void showMaxSessionDialog() {
        BottomSheetDialog bt = new BottomSheetDialog(activity(), R.style.BottomSheetDialogTheme);
        View view;
        bt.setDismissWithAnimation(true);
        bt.setContentView(view = Uis.inflate(activity(), R.layout.view_schedule_time));
        bt.getWindow().findViewById(R.id.design_bottom_sheet).setBackgroundDrawable(new ColorDrawable(Color.TRANSPARENT));
        bt.show();
        view.animate().translationY(50);

        TextView tv = view.findViewById(R.id.dialog_title);
        SeekBar s1 = view.findViewById(R.id.seekBar);
        SeekBar s2 = view.findViewById(R.id.seekBar2);
        TextView t1 = view.findViewById(R.id.tvS1);
        TextView t2 = view.findViewById(R.id.tvS2);
        Button button = view.findViewById(R.id.button);

        tv.setText("配置最大节次");
        s2.setVisibility(View.GONE);
        t2.setVisibility(View.GONE);


        s1.setMax(20);
        s1.setProgress(schedule.maxSession);
        t1.setText(schedule.maxSession + "节");
        s1.setOnSeekBarChangeListener(new MySeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                t1.setText(progress + "节");
            }
        });
        button.setOnClickListener(v -> {
            if (s1.getProgress() == 0) {
                Toasts.toast(getString(R.string.param_is_illgal));
            } else {
                schedule.maxSession = s1.getProgress();
                globalT.scheduleMaxSessionSubtitle.setText(s1.getProgress() + "节");
                scheduleViewModel.updateSchedule(schedule);
                bt.dismiss();
            }
        });
    }


    /**
     * 课程卡片透明度
     */
    @SuppressLint("SetTextI18n")
    private void showCourseItemAlphaDialog() {
        BottomSheetDialog bt = new BottomSheetDialog(activity(), R.style.BottomSheetDialogTheme);
        View view;
        bt.setDismissWithAnimation(true);
        bt.setContentView(view = Uis.inflate(activity(), R.layout.view_schedule_time));
        bt.getWindow().findViewById(R.id.design_bottom_sheet).setBackgroundDrawable(new ColorDrawable(Color.TRANSPARENT));
        bt.show();
        view.animate().translationY(50);

        TextView tv = view.findViewById(R.id.dialog_title);
        SeekBar s1 = view.findViewById(R.id.seekBar);
        SeekBar s2 = view.findViewById(R.id.seekBar2);
        TextView t1 = view.findViewById(R.id.tvS1);
        TextView t2 = view.findViewById(R.id.tvS2);
        Button button = view.findViewById(R.id.button);

        tv.setText("配置透明度");
        s2.setVisibility(View.GONE);
        t2.setVisibility(View.GONE);


        s1.setMax(10);
        s1.setProgress(schedule.alphaForCourseItem);
        t1.setText("L" + schedule.alphaForCourseItem);
        s1.setOnSeekBarChangeListener(new MySeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                t1.setText("L" + progress);
            }
        });
        button.setOnClickListener(v -> {
            schedule.alphaForCourseItem = s1.getProgress();
            globalT.scheduleCourseAlphaSubTitle.setText("L" + s1.getProgress());
            scheduleViewModel.updateSchedule(schedule);
            bt.dismiss();
        });
    }


    /**
     * 时间配置
     */
    @SuppressLint("StringFormatMatches")
    private void showTimeConfigDialog() {
        BottomSheetDialog bt = new BottomSheetDialog(activity(), R.style.BottomSheetDialogTheme);
        View view;
        bt.setDismissWithAnimation(true);
        bt.setContentView(view = Uis.inflate(activity(), R.layout.view_schedule_time));
        bt.getWindow().findViewById(R.id.design_bottom_sheet).setBackgroundDrawable(new ColorDrawable(Color.TRANSPARENT));
        bt.show();
        view.animate().translationY(50);

        SeekBar s1 = view.findViewById(R.id.seekBar);
        SeekBar s2 = view.findViewById(R.id.seekBar2);
        TextView t1 = view.findViewById(R.id.tvS1);
        TextView t2 = view.findViewById(R.id.tvS2);
        Button button = view.findViewById(R.id.button);

        final int[] mTotalWeek = new int[1];
        final int[] mCurWeek = new int[1];


        //初始化
        s1.setMax(30);
        s1.setProgress(schedule.totalWeek);
        s2.setMax(schedule.totalWeek);
        s2.setProgress(schedule.curWeek());
        mTotalWeek[0] = schedule.totalWeek;
        mCurWeek[0] = schedule.curWeek();
        t1.setText(getString(R.string.total_week, schedule.totalWeek));
        t2.setText(getString(R.string.current_week, schedule.curWeek()));

        //监听
        s1.setOnSeekBarChangeListener(new MySeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                s2.setMax(progress);
                t1.setText(getString(R.string.total_week, String.valueOf(progress)));
                mTotalWeek[0] = progress;
            }
        });

        s2.setOnSeekBarChangeListener(new MySeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                t2.setText(getString(R.string.current_week, String.valueOf(progress)));
                mCurWeek[0] = progress;
            }
        });

        button.setOnClickListener(v -> {
            if (mTotalWeek[0] == 0 || mCurWeek[0] == 0) {
                Toasts.toast(getString(R.string.param_is_illgal));
            } else {
                schedule.totalWeek = mTotalWeek[0];
                schedule.termStartDate = Dates.getTermStartDate(mCurWeek[0]);
                globalT.scheduleWeekSubtitle.setText(t1.getText().toString() + t2.getText().toString());
                scheduleViewModel.updateSchedule(schedule);
                bt.dismiss();
            }
        });
    }


    /**
     * 开始选择图片
     */
    private void startChoosePicture() {
        Intent intent = new Intent(Intent.ACTION_PICK, MediaStore.Images.Media.EXTERNAL_CONTENT_URI);
        startActivityForResult(intent, PICK_PHOTO);
    }


    @Override
    public void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        if (data == null) {
            return;
        }
        if (requestCode == PICK_PHOTO) {
            if (resultCode == RESULT_OK) {
                handleImage(data);
            }
        }
    }


    /**
     * 处理图片
     *
     * @param data
     */
    @SuppressLint("CheckResult")
    private void handleImage(Intent data) {
        Uri uri = data.getData();
        ContentResolver cr = activity().getContentResolver();
        Bitmap bitmap;
        try {
            bitmap = BitmapFactory.decodeStream(cr.openInputStream(uri));
            Toasts.toast(getString(R.string.update_success));
            schedule.imageUrl = Files.getFilePath(activity(), uri);
            scheduleViewModel.updateSchedule(schedule);
            Glide.with(activity()).load(new File(schedule.imageUrl)).into(globalT.backgroundPic);
        } catch (FileNotFoundException e) {
            e.printStackTrace();
            Toasts.toast(getString(R.string.pic_choose_fail));
        }
    }

}

File: app/src/main/java/cn/surine/schedulex/ui/schedule_config/ScheduleConfigFragment.kt
package cn.surine.schedulex.ui.schedule_config

import android.Manifest
import android.annotation.SuppressLint
import android.app.Activity.RESULT_OK
import android.content.Intent
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.os.Build
import android.os.Bundle
import android.provider.MediaStore
import android.view.View
import android.widget.Button
import android.widget.SeekBar
import android.widget.TextView
import cn.surine.schedulex.R
import cn.surine.schedulex.app_base.VmManager
import cn.surine.schedulex.base.Constants
import cn.surine.schedulex.base.controller.BaseBindingFragment
import cn.surine.schedulex.base.utils.*
import cn.surine.schedulex.data.entity.Schedule
import cn.surine.schedulex.databinding.FragmentScheduleConfigBinding
import cn.surine.schedulex.ui.course.CourseViewModel
import cn.surine.schedulex.ui.schedule.ScheduleViewModel
import cn.surine.schedulex.ui.schedule_list.ScheduleListFragment
import cn.surine.schedulex.ui.schedule_list.ScheduleListFragment.Companion.CHANGE_BACKGROUND
import cn.surine.schedulex.ui.schedule_list.ScheduleListFragment.Companion.CHANGE_COURSE_ITEM_HEIGHT
import cn.surine.schedulex.ui.schedule_list.ScheduleListFragment.Companion.CHANGE_SCHEDULE_NAME
import cn.surine.schedulex.ui.schedule_list.ScheduleListFragment.Companion.CHANGE_WEEK_INFO
import cn.surine.schedulex.ui.schedule_list.ScheduleListFragment.Companion.PICK_PHOTO
import cn.surine.schedulex.ui.schedule_list.ScheduleListFragment.Companion.SCHEDULE_ID
import cn.surine.schedulex.ui.timetable_list.TimeTableViewModel
import cn.surine.schedulex.ui.view.custom.helper.CommonDialogs
import com.bumptech.glide.Glide
import com.google.android.material.bottomsheet.BottomSheetDialog
import com.peanut.sdk.miuidialog.MIUIDialog
import com.tbruyelle.rxpermissions2.RxPermissions
import kotlinx.android.synthetic.main.fragment_schedule_config.*
import kotlinx.android.synthetic.main.view_single_slider_ui.view.*
import java.io.File

/**
 * Intro：
 *
 * @author sunliwei
 * @date 2020/6/24 16:28
 */
class ScheduleConfigFragment : BaseBindingFragment<FragmentScheduleConfigBinding>() {

    private lateinit var scheduleViewModel: ScheduleViewModel
    private lateinit var courseViewModel: CourseViewModel
    private lateinit var timetableViewModel: TimeTableViewModel

    private var scheduleId: Int = -1
    private lateinit var schedule: Schedule
    private var mSettingItemTag: Int = -1

    override fun layoutId(): Int = R.layout.fragment_schedule_config
    override fun onInit(t: FragmentScheduleConfigBinding) {
        VmManager(this).apply {
            scheduleViewModel = vmSchedule
            courseViewModel = vmCourse
            timetableViewModel = vmTimetable
        }
        scheduleId = arguments?.getInt(SCHEDULE_ID) ?: -1
        if (scheduleId.toLong() == Prefs.getLong(Constants.CUR_SCHEDULE, -1)) {
            deleteSchedule.visibility = View.GONE
        }
        schedule = scheduleViewModel.getScheduleById(scheduleId.toLong())
        t.data = schedule
        //捷径
        mSettingItemTag = arguments?.getInt(ScheduleListFragment.FUNCTION_TAG, -1) ?: -1
        if (mSettingItemTag != -1) {
            when (mSettingItemTag) {
                CHANGE_WEEK_INFO -> showTimeConfigDialog()
                CHANGE_COURSE_ITEM_HEIGHT -> showItemHeightDialog()
                CHANGE_BACKGROUND -> chooseBackgroundPicture()
                CHANGE_SCHEDULE_NAME -> modifyScheduleName()
            }
            mSettingItemTag = -1
        }


        //删除
        deleteSchedule.setOnClickListener {
            CommonDialogs.getCommonDialog(activity(), getString(R.string.warning), getString(R.string.delete_schedule_dialog_msg), okCall = {
                scheduleViewModel.deleteScheduleById(scheduleId.toLong())
                courseViewModel.deleteCourseByScheduleId(scheduleId.toLong())
                Toasts.toast(getString(R.string.schedule_is_delete))
                Navigations.close(this@ScheduleConfigFragment)
            })
        }

        //课表名
        scheduleNameItem.setOnClickListener { modifyScheduleName() }
        //周信息
        scheduleWeekInfoItem.setOnClickListener { showTimeConfigDialog() }
        //时间表
        scheduleTimeTableItem.setOnClickListener {
            Navigations.open(this, R.id.action_scheduleConfigFragment_to_timeTableListFragment, Bundle().apply {
                putLong(SCHEDULE_ID, scheduleId.toLong())
            })
        }
        timeTableSubtitle.text = timetableViewModel.getTimTableById(schedule.timeTableId)?.name
                ?: "No Data"
        //背景图
        scheduleBackgroundItem.setOnClickListener { chooseBackgroundPicture() }
        scheduleBackgroundItem.setOnLongClickListener {
            scheduleViewModel.updateSchedule(schedule.apply { imageUrl = "" })
            true
        }
        if (schedule.imageUrl != null && schedule.imageUrl!!.isNotEmpty()) {
            Glide.with(activity()).load(File(schedule.imageUrl)).into(backgroundPic)
        }
        //颜色适配
        switchs.isChecked = schedule.lightText
        schedulePaletteItem.setOnClickListener {
            schedule.lightText = !schedule.lightText
            scheduleViewModel.updateSchedule(schedule)
            switchs.isChecked = schedule.lightText
            paletteColorSubTitle.setText(if (schedule.lightText) R.string.white_txt else R.string.black_txt)
        }
        //设置是否展示周末开关
        showWeekSwitchs.isChecked = schedule.isShowWeekend
        scheduleShowWeekItem.setOnClickListener {
            schedule.isShowWeekend = !schedule.isShowWeekend
            scheduleViewModel.updateSchedule(schedule)
            showWeekSwitchs.isChecked = schedule.isShowWeekend
            showWeekSubTitle.setText(if (schedule.isShowWeekend) R.string.show_weekend else R.string.not_show_weekend)
        }
        //打开透明度配置窗口
        scheduleCourseAlphaItem.setOnClickListener { showCourseItemAlphaDialog() }
        //配置最大节次
        scheduleMaxSessionItem.setOnClickListener { showMaxSessionDialog() }
        //配置课表最大高度
        scheduleCourseItemHeightItem.setOnClickListener { showItemHeightDialog() }
        //是否显示时间表
        showTimesSwitch.isChecked = schedule.isShowTime
        t.scheduleShowTimeItem.setOnClickListener {
            schedule.isShowTime = !schedule.isShowTime
            scheduleViewModel.updateSchedule(schedule)
            showTimesSwitch.isChecked = schedule.isShowTime
            scheduleShowTimeItemSubTitle.setText(if (schedule.isShowTime) R.string.show_time else R.string.not_show_timetable)
        }
        //课程卡片主题
        changeCourseCardUi.setOnClickListener {
            Navigations.open(this, R.id.action_scheduleConfigFragment_to_paletteFragment, Bundle().apply {
                putInt(SCHEDULE_ID, scheduleId)
            })
        }
        changeCharLimit.setOnClickListener {
            showChangeCharLimitDialog()
        }
        //导出
        export.setOnClickListener {
            Navigations.open(this, R.id.action_scheduleConfigFragment_to_scheduleDataExport, Bundle().apply {
                putInt(SCHEDULE_ID, scheduleId)
            })
        }
    }

    private fun showChangeCharLimitDialog() {
        MIUIDialog(activity()).show {
            customView(R.layout.view_single_slider_ui) {
                it.single_slider_view_title.text = "调整数值"
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                    it.singleSeekBar.min = 1
                }
                it.singleSeekBar.max = 20
                it.singleSeekBar.progress = schedule.maxHideCharLimit
                it.singleText.text = "${schedule.maxHideCharLimit}"
                it.singleSeekBar.setOnSeekBarChangeListener(object : MySeekBarChangeListener() {
                    override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
                        it.singleText.text = "$progress"
                    }
                })
                it.button.setOnClickListener { _->
                    schedule.maxHideCharLimit = it.singleSeekBar.progress
                    scheduleViewModel.updateSchedule(schedule)
                    dismiss()
                }
            }
        }
    }

    private fun modifyScheduleName() {
        MIUIDialog(activity()).show {
            title(text = "编辑课表名称")
            input(prefill = schedule.name) { text, _ ->
                schedule.name = text.toString()
                this@ScheduleConfigFragment.courseNameSubTitle.text = text
                Toasts.toast(getString(R.string.update_success))
                scheduleViewModel.updateSchedule(schedule)
            }
            positiveButton(text = "保存") { }
            negativeButton(text = "取消") { }
        }
    }

    @SuppressLint("CheckResult")
    private fun chooseBackgroundPicture() {
        RxPermissions(activity()).request(Manifest.permission.READ_EXTERNAL_STORAGE).subscribe {
            if (it) {
                startActivityForResult(Intent(Intent.ACTION_PICK, MediaStore.Images.Media.EXTERNAL_CONTENT_URI), PICK_PHOTO)
            } else Toasts.toast(getString(R.string.permission_is_denied))
        }
    }

    @SuppressLint("SetTextI18n")
    private fun showItemHeightDialog() {
        BottomSheetDialog(activity(), R.style.BottomSheetDialogTheme).apply {
            dismissWithAnimation = true
            val view: View = Uis.inflate(activity(), R.layout.view_schedule_time)
            setContentView(view)
            window?.findViewById<View>(R.id.design_bottom_sheet)?.setBackgroundDrawable(ColorDrawable(Color.TRANSPARENT))
            show()
            view.animate().translationY(50F)
            val tv = view.findViewById<TextView>(R.id.dialog_title)
            val s1 = view.findViewById<SeekBar>(R.id.seekBar)
            val s2 = view.findViewById<SeekBar>(R.id.seekBar2)
            val t1 = view.findViewById<TextView>(R.id.tvS1)
            val t2 = view.findViewById<TextView>(R.id.tvS2)
            val button = view.findViewById<Button>(R.id.button)
            tv.text = "配置课程格子高度"
            s2.visibility = View.GONE
            t2.visibility = View.GONE
            s1.max = 150
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                s1.min = 30
            }
            s1.progress = schedule.itemHeight
            t1.text = "${schedule.itemHeight} dp"
            s1.setOnSeekBarChangeListener(object : MySeekBarChangeListener() {
                override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
                    t1.text = "$progress dp"
                }
            })
            button.setOnClickListener {
                if (s1.progress < 30) {
                    Toasts.toast(getString(R.string.cant_small_than_30))
                } else {
                    schedule.itemHeight = s1.progress
                    //6月24 @bug：在Dialog上下文中访问fragment的资源变量会产生空指针
                    this@ScheduleConfigFragment.scheduleCourseItemHeightSubtitle.text = "${s1.progress} px"
                    scheduleViewModel.updateSchedule(schedule)
                    dismiss()
                }
            }
        }
    }


    @SuppressLint("SetTextI18n")
    private fun showMaxSessionDialog() {
        BottomSheetDialog(activity(), R.style.BottomSheetDialogTheme).apply {
            dismissWithAnimation = true
            val view: View = Uis.inflate(activity(), R.layout.view_schedule_time)
            setContentView(view)
            window?.findViewById<View>(R.id.design_bottom_sheet)?.setBackgroundDrawable(ColorDrawable(Color.TRANSPARENT))
            show()
            view.animate().translationY(50F)
            val tv = view.findViewById<TextView>(R.id.dialog_title)
            val s1 = view.findViewById<SeekBar>(R.id.seekBar)
            val s2 = view.findViewById<SeekBar>(R.id.seekBar2)
            val t1 = view.findViewById<TextView>(R.id.tvS1)
            val t2 = view.findViewById<TextView>(R.id.tvS2)
            val button = view.findViewById<Button>(R.id.button)
            tv.text = "配置最大节次"
            s2.visibility = View.GONE
            t2.visibility = View.GONE
            s1.max = 20
            s1.progress = schedule.maxSession
            t1.text = "${schedule.maxSession} 节"
            s1.setOnSeekBarChangeListener(object : MySeekBarChangeListener() {
                override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
                    t1.text = "$progress 节"
                }
            })
            button.setOnClickListener {
                if (s1.progress == 0) {
                    Toasts.toast(getString(R.string.param_is_illgal))
                } else {
                    schedule.maxSession = s1.progress
                    this@ScheduleConfigFragment.scheduleMaxSessionSubtitle.text = "${s1.progress} 节"
                    scheduleViewModel.updateSchedule(schedule)
                    dismiss()
                }
            }
        }
    }

    @SuppressLint("SetTextI18n")
    private fun showCourseItemAlphaDialog() {
        BottomSheetDialog(activity(), R.style.BottomSheetDialogTheme).apply {
            dismissWithAnimation = true
            val view: View = Uis.inflate(activity(), R.layout.view_schedule_time)
            setContentView(view)
            window?.findViewById<View>(R.id.design_bottom_sheet)?.setBackgroundDrawable(ColorDrawable(Color.TRANSPARENT))
            show()
            view.animate().translationY(50F)
            val tv = view.findViewById<TextView>(R.id.dialog_title)
            val s1 = view.findViewById<SeekBar>(R.id.seekBar)
            val s2 = view.findViewById<SeekBar>(R.id.seekBar2)
            val t1 = view.findViewById<TextView>(R.id.tvS1)
            val t2 = view.findViewById<TextView>(R.id.tvS2)
            val button = view.findViewById<Button>(R.id.button)
            tv.text = "配置透明度"
            s2.visibility = View.GONE
            t2.visibility = View.GONE
            s1.max = 10
            s1.progress = schedule.alphaForCourseItem
            t1.text = "L ${schedule.alphaForCourseItem}"
            s1.setOnSeekBarChangeListener(object : MySeekBarChangeListener() {
                override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
                    t1.text = "L $progress"
                }
            })
            button.setOnClickListener {
                schedule.alphaForCourseItem = s1.progress
                this@ScheduleConfigFragment.scheduleCourseAlphaSubTitle.text = "L ${s1.progress}"
                scheduleViewModel.updateSchedule(schedule)
                dismiss()
            }
        }
    }


    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        data ?: return
        if (requestCode == PICK_PHOTO) {
            if (resultCode == RESULT_OK) {
                handleImage(data)
            }
        }
    }

    private fun handleImage(data: Intent) {
        try {
            val uri = data.data
            Toasts.toast(getString(R.string.update_success))
            schedule.imageUrl = Files.getFilePath(activity(), uri)
            scheduleViewModel.updateSchedule(schedule)
            Glide.with(activity()).load(File(schedule.imageUrl)).into(backgroundPic)
        } catch (e: Exception) {
            Toasts.toast(getString(R.string.pic_choose_fail))
        }
    }

    @SuppressLint("StringFormatMatches", "SetTextI18n")
    private fun showTimeConfigDialog() {
        BottomSheetDialog(activity(), R.style.BottomSheetDialogTheme).apply {
            dismissWithAnimation = true
            val view: View = Uis.inflate(activity(), R.layout.view_schedule_time)
            setContentView(view)
            window?.findViewById<View>(R.id.design_bottom_sheet)?.setBackgroundDrawable(ColorDrawable(Color.TRANSPARENT))
            show()
            view.animate().translationY(50F)
            val s1 = view.findViewById<SeekBar>(R.id.seekBar)
            val s2 = view.findViewById<SeekBar>(R.id.seekBar2)
            val t1 = view.findViewById<TextView>(R.id.tvS1)
            val t2 = view.findViewById<TextView>(R.id.tvS2)
            val button = view.findViewById<Button>(R.id.button)

            var mTotalWeek: Int = schedule.totalWeek
            var mCurWeek = schedule.curWeek()
            s1.max = 30
            s1.progress = schedule.totalWeek
            s2.max = schedule.totalWeek
            s2.progress = schedule.curWeek()
            t1.text = getString(R.string.total_week, schedule.totalWeek)
            t2.text = getString(R.string.current_week, schedule.curWeek())

            s1.setOnSeekBarChangeListener(object : MySeekBarChangeListener() {
                override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
                    s2.max = progress
                    t1.text = getString(R.string.total_week, progress.toString())
                    mTotalWeek = progress
                }
            })

            s2.setOnSeekBarChangeListener(object : MySeekBarChangeListener() {
                override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
                    t2.text = getString(R.string.current_week, progress.toString())
                    mCurWeek = progress
                }
            })

            button.setOnClickListener {
                if (mTotalWeek == 0 || mCurWeek == 0) {
                    Toasts.toast(getString(R.string.param_is_illgal))
                } else {
                    schedule.totalWeek = mTotalWeek
                    schedule.termStartDate = Dates.getTermStartDate(mCurWeek)
                    this@ScheduleConfigFragment.scheduleWeekSubtitle.text = t1.text.toString() + t2.text.toString()
                    scheduleViewModel.updateSchedule(schedule)
                    dismiss()
                }
            }
        }
    }
}